optax/contrib/reduce_on_plateau.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2023 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Reduce Learning Rate on Plateau callback.
+
+This callback monitors a quantity and if no improvement is seen for a 'patience'
+number of epochs, the learning rate is reduced by a factor of 'reduce_factor'.
+Optionally, a cooldown period can be specified during which the learning rate
+will not be reduced.
+"""
+from typing import NamedTuple
+
+import jax
+import jax.numpy as jnp
+
+from optax._src import base
+
+
+class ReduceLROnPlateauState(NamedTuple):
+  """State for the ReduceLROnPlateau callback."""
+  reduce_factor: float
+  patience: int
+  min_improvement: float
+  best_loss: float
+  plateau_count: int
+  lr: float
+  cooldown_counter: int
+  cooldown:int
+
+
+def reduce_on_plateau(
+  reduce_factor: float,
+  patience: int,
+  min_improvement:float,
+  cooldown:int
+) -> base.GradientTransformationExtraArgs:
+  """ Args:
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning 
```

Comment:
Why are these parameters in terms of number of epochs? Wouldn't it be more accurate to talk about number of iterations? It seems to me that the method has now knowledge of epochs, but keeps track of iterations instead